### PR TITLE
Reduce Rpc Continuation Queue Locking

### PR DESCRIPTION
**Pull request by YulerB**
## Proposed Changes

- Updated RpcContinuationQueue.cs.
- Removed locking and replaced with an atomic operation.
- This reduces memory and mean time to execute.
- Also making the variable that is being locked private.
- Also removes the chance of a null exception in code, where the output of Next() is not checked.
- Added tests.
- Benchmark test code and outcome added to comments below.

## Types of Changes

What types of changes does your code introduce to this project?

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [X] Performance Update  

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)
- [X] Any dependent changes have been merged and published in related repositories

## Further Comments

```
BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i5-6200U CPU 2.30GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
Frequency=2343750 Hz, Resolution=426.6667 ns, Timer=TSC
  [Host]     : .NET Framework 4.6.1 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3132.0  [AttachedDebugger]
  Job-JWVBYN : .NET Framework 4.6.1 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3132.0
Runtime=Clr  IsBaseline=True  
Method     Mean        Error           StdDev       Scaled      Rank       Gen 0        Allocated
RpcNew    26.94 ns    0.3565 ns    0.3334 ns   1.00          1             0.0076      12 B
RpcOld      73.05 ns    2.7595 ns    3.1779 ns   1.00          1             0.0178      28 B
```

Bench mark code:
[bench.cs.txt](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/2364570/bench.cs.txt)

### CODE REVIEW

**Comment by kjnilsson** [comment]
This looks good to me but as usual with these type of changes it would be good to get a couple of eyes on it.

File: projects/client/Unit/src/unit/TestRpcContinuationQueue.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+//---------------------------------------------------------------------------
+
+using NUnit.Framework;
+using RabbitMQ.Client.Impl;
+using System;
+
+namespace RabbitMQ.Client.Unit
+{
+    [TestFixture]
+    public class TestRpcContinuationQueue {
```
Comment (lukebakken) [inline projects/client/Unit/src/unit/TestRpcContinuationQueue.cs]: Please ensure that the code formatting follows the rest of the project's style by placing the opening curly on a new line.

File: projects/client/Unit/src/unit/TestRpcContinuationQueue.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        [Test]
+        public void TestRpcContinuationQueueEnqueueAndRelease()
+        {
+            RpcContinuationQueue queue = new RpcContinuationQueue();
+            var inputContinuation = new SimpleBlockingRpcContinuation();
+            queue.Enqueue(inputContinuation);
+            var outputContinuation = queue.Next();
+            Assert.AreEqual(outputContinuation, inputContinuation);
+        }
```
Comment (lukebakken) [inline projects/client/Unit/src/unit/TestRpcContinuationQueue.cs]: Insert a blank line here.

File: projects/client/Unit/src/unit/TestRpcContinuationQueue.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        public void TestRpcContinuationQueueEnqueueAndRelease2()
+        {
+            RpcContinuationQueue queue = new RpcContinuationQueue();
+            var inputContinuation = new SimpleBlockingRpcContinuation();
+            queue.Enqueue(inputContinuation);
+            var outputContinuation = queue.Next();
+            Assert.AreEqual(outputContinuation, inputContinuation);
+            var outputContinuation1 = queue.Next();
+            Assert.AreNotEqual(outputContinuation1, inputContinuation);
+        }
```
Comment (lukebakken) [inline projects/client/Unit/src/unit/TestRpcContinuationQueue.cs]: Insert a blank line here.

File: projects/client/Unit/src/unit/TestRpcContinuationQueue.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            RpcContinuationQueue queue = new RpcContinuationQueue();
+            var inputContinuation = new SimpleBlockingRpcContinuation();
+            var inputContinuation1 = new SimpleBlockingRpcContinuation();
+            queue.Enqueue(inputContinuation);
+            Assert.Throws(typeof(NotSupportedException), () => {
+                queue.Enqueue(inputContinuation1);
+            });
+        }
+
```
Comment (lukebakken) [inline projects/client/Unit/src/unit/TestRpcContinuationQueue.cs]: Please remove extra blank lines.

File: projects/client/Unit/src/unit/TestRpcContinuationQueue.cs (hunk 3 above)
Comment (YulerB) [reply projects/client/Unit/src/unit/TestRpcContinuationQueue.cs]: @lukebakken, please review if line breaks are now correct?

**Comment by michaelklishin** [comment]
Thank you @YulerB.

**Pull request merged**